### DISCUSSION TITLE
Telemetry extension

### DISCUSSION THREAD

**Discussion by gagarinlg** [crsf-wg/crsf#14] [Enhancements]
following a discussion on the EdgeTX discord, this discussion shall be used as a starting point for a telemetry üprotocol rewrite

Basic ideas:
- sensor discovery on the RX side shall be started by a "discover sensors" command from the TX
- sensors deliver: name, type, value range, unit, data size, scaling (thus removing the need to transport floats)
- senor refresh rates are configurable wvia the LUA script and stored in the sensors
- packet transmission over the air, when the sensor IDs are consecutive
  - ID, number of values, values 
- sensors are conencted as a chain, the sensor closest to the RX has ID0, next sensor in the chain ID 1, ...
  - the sensors form a daisy chain, where one sensor is connected to the next. The IDs are the position in the chain
  - that way, when using the same telemetry port and the same sensor order in the chain, the sensor IDs are reproducable over multiple models
- RX can have multiple sensor chains, sensor chains are numberd by prefixing  a few bits to the sensor IDs
- RX polls sensors, when there is never an answer within the first x (5?) seconds after power up, the sensor will not be polled again until the next power up
  - optionally add commands from external devices to the RX to add sensors
- FCs have to manage sensor IDs on there own, no need to implement chains there
  - sensor chains are primarily for PWM receivers used without a FC, so that sensors can be connected and sensor data be transmitted

TBD:
- physical interface
  - 3 pin, +, - , signal ?
  - speed
  - ...
- CRSF command extensions

Reasoning:
- this system would allow for a flexible telemtry system, where new sensor types do not trigger code changes
- for usages where the same model storage in the TX is used for multiple "RC things", the chain system creates reproducable sensor IDs
- adding multiple different telemetry sensor protocol types on the RX would not require resuing existing sensor types

**Comment by wimalopaan** [comment]
Do you suggest to physically connect the sensors as a chain? This looks like FlySky/IBus (but IBus has only 15 sensor IDs).

**Comment by froqstar** [comment]
> senor refresh rates are configurable wvia the LUA script and stored in the sensors

I fear that configuring them as refresh rates per sensor may be too complicated for users, as you always need to take (remaining) telemetry bandwidth into account, which with variable telemetry item size is non-trivial.
Maybe defining priorities which automatically assigns refresh rates to the sensors depending on the available bandwidth could be easier? 

In the end we have to distinguish between protocol and user interface, as long as the protocol is flexible enough the interface can hide most of the complexity.

**Comment by snakescb** [comment]
Amazing proposal indeed, one stable but flexible and future proof telemetry stack, on or above the level what current expensive radio systems offer!

- I would rather combine sensor ID's from receiver number (maybe 2 bits) followed by 6 bits sensor number  or so. This will allow more than one receiver to send telemetry information, which is a common feature in RC links today. If sensors are connected to different "chains" or pins on the RX, the receiver can figure out the sequence itself 
- Refresh rate could be configured per receiver, or globally. Probably not per sensor, not sure how benefitial this is relative to complexity of implementing and configuring it.
- During Sensor discovery, sensors should also inform about the amount of databytes they need for their information update. A simple temperature sensor may be ok with a single databyte, where a GPS may require up to 8. To make most use of the available bandwidth, the data bytes required per update should be flexible
- Also, data type (signed/unsigned) and decimal position should be flexible and configured during discovery
- I like that sensor name and unit in text representation are flexible, that's super neat!

**Comment by mmosca** [comment]
I agree that refresh rates should not be delegated to the sensors. There is usually enough bandwidth between sensor/fc/rx  to transmit data as needed.

RX has limited bandwidth and can manage that better as it is aware of how often it can send data.

It is unclear to me how a chain can result in fixed ids between different models, specially when sharing a model across multiple aircraft with different sensor combinations. How are ids allocated?

Depending on how we allocate ids, chains can help disambiguate duplicate sensors. (multiple temperature or rpm sensors, as an example).

We need to find a balance between the mess that we have configuring other systems and flexibility, for example, Futaba has users assigning "slots" to different telemetry sensors. Assigning a slot means changing settings in the radio  AND in the sensor (let's avoid this, if we can).

A physical bus or chain adds a lot of complexity when you have independent sensors that need to coordinate access to it and will add a lot of complexity. This should probably be handled separately, and the protocol discussion should concentrate on that being handled somewhere else.

FCs already support direct input from multiple sensors, over different pins/protocols/buses and can consolidate the info without needing to implement this chain or bus.

RXs can also have dedicated sensor inputs, in addition to this new physical sensor telemetry protocol.

**Reply by snakescb** [reply]
Futaba is a mess. One balance could be, that one receiver pin can receive data from one sensor only, rather than a whole bus. The sensor can provide more than one telemetry item still (what an Flight Controller or a PowerBox is doing). How to manage different RC things with the same transmitter model when they may have different sensors, this is a tough one. Existing systems do not really have this functionality.

**Reply by mmosca** [reply]
Yes. 100% let's avoid the mess that Futaba is. I think the bus/chain discussion is somewhat separate/parallel from the sensor extensibility.

There are a few downsides to this bus/chain thing:
* Not required to add sensor flexibility. FCs can quickly add new sensors without the "BUS/Chain"  being implemented.
* New protocol with no sensors on the market
  * You will need a mcu to convert your sensor input into this protocol
  * Extra work for FCs/Powerbox/extra devices to add new telemetry input protocol
  * Harder for FC devs to add/validate support
     * We can probably simulate sensors, but there will be teething pains when buggy hardware makes it to market.

We may end up in a Chicken and egg situation. No FC support because there are no sensors, and no sensors because there are no support. Specially true, if you are expecting to closed systems like powerbox to support these sensors.

**Reply by snakescb** [reply]
We will probably not see new sensors with our CRSF telemetry protocol. What happens is, that the FC's or receivers will understand existing  sensor language, and translate to CRSF. I would love to add Jeti ExBus support to ExpressLRS receivers. They can be connected then to anything (and there is a lot) that can deal with ExBus (like PowerBoxes in Jet planes), and the receiver can understand and interprete the data. But we need to send it to the transmitter, and display it in the handheld after

**Reply by gagarinlg** [reply]
Let's split the answer:

- there will be stand alone senors available as soon as it is possible to connect them and transmit the information
- the whole sensor chain thing is more for stand alone receivers that are used without an FC, so primarily PWM receivers
  - FCs can make there own ID scheme, no need to implement the physical part, only the logical protocol level part, when walking to the RX

**Reply by mmosca** [reply]
I think this is describing 2 separate concerns:

1. Sensor discovery and extensibility between rx and tx.
2. New physical way to connect sensors to RX and sensor protocol.

2 is not needed for 1. Existing crsf sensor frames can and should be converted to new extensible sensor style by the RX.

2 makes more sense with 1 in place.

**Reply by gagarinlg** [reply]
I would like to have this as one effort. Then we can create a consistent telemetry solution instead of patching thing into what is currently there and 3 months later redo everything, because the patches are bad for code maintenance and we probably forget something.

**Reply by mmosca** [reply]
> I would like to have this as one effort. Then we can create a consistent telemetry solution instead of patching thing into what is currently there and 3 months later redo everything, because the patches are bad for code maintenance and we probably forget something.

We probably need to find some balance. Whatever we do for 1, should make 2 easier without requiring big changes, but it should also be usable for existing protocols, including current crsf implementation.

As an exercise,we have already described a few types of sensor that we will need frames for.
1. Simple single value sensor: label, number of decimal places, value, unit, number of bits? Integer based + decimal places
2. Complex sensor: gps coordinates: Can it be treated as multiple instances of 1 (Longitue + latitude sensors, etc) or does it need its own type to be treated as a single unit?
3. What are the units used for? Arbitrary text for display only, or can they be used for conversion on the handset side and need an official catalog and binary code? (m is 1, m/s is 2, feet/s is 3, etc...)

Anything else, sensor wise I have missed?

**Reply by mmosca** [reply]
Is data size/bandwidth a concern on the wire, or just over the air (OTA)? Can we simplify data exchange by supporting 16/32bit integers and floats for exchange?

**Reply by gagarinlg** [reply]
OTA is the issue
Using a scaling value ate sensor discovery and the only transmit integers should be possible

**Reply by CapnBry** [reply]
> Is data size/bandwidth a concern on the wire, or just over the air (OTA)? Can we simplify data exchange by supporting 16/32bit integers and floats for exchange?

🤣 It is a major major concern. Encoding **must** be used to reduce the data size as much as possible by the sensor (or the intermediary if the sensor is another protocol). Floats are pretty much right out.

**Comment by wimalopaan** [comment]
Well, bringing a new protocoll to the zoo of protocols does not help in my opinion. 

CRSF as it is, is the de-facto standard as a full-duplex proto between receiver and FC (concentrator). CRSF has soem shortcomings (see #3, #4 and #12). I think we should concentrate to solve these issues instead of inventing a new half-duplex telem protocol.

FCs implement the well known telem protocols like S.Port, SBus2, IBus, Hott. The bottleneck is that CRSF from RX->TX does not support all sensor types. This shoulb be solved.

On the other hand, as more and more PWM-RX come into play, it would be cool, if ELRS could support the established telem protos as well (like S.Port, SBus2, IBus, Hott, ...), maybe with 2 or more UARTS with different protos.
But also here the bottleneck is that CRSF (on the TX side) does not support all sensor types.

And last: who would implement the new telem. proto? Maybe OpenXSensor, but not much more.
And last 2: inventing the new telem proto requires the shortcomings of RX->TX->Handset to be solved, that we have a lack of sensortypes.

**Reply by snakescb** [reply]
> FCs implement the well known telem protocols like S.Port, SBus2, IBus, Hott. The bottleneck is that CRSF from RX->TX does not support all sensor types. This shoulb be solved.

Fully agree! We dont need new sensor protocols. We need flexible transport rx->tx->handset

**Reply by gagarinlg** [reply]
I disagree.
Implementing more and more sensor protocols into theELRS receivers, because there is another user using another brand of sensors is just lost time.
There are many users never using FCs, me included. So jsut looking from an FC perspective does not really help. Nobody is proposing to remove those protocols that are already supported.

**Reply by mmosca** [reply]
This is always a concern:
![standards](https://github.com/crsf-wg/crsf/assets/23555060/5612dddd-4500-4e8f-93ee-e98ec013694f)

**Reply by gagarinlg** [reply]
Of course

**Reply by wimalopaan** [reply]
> I disagree. Implementing more and more sensor protocols into theELRS receivers, because there is another user using another brand of sensors is just lost time. There are many users never using FCs, me included. So jsut looking from an FC perspective does not really help. Nobody is proposing to remove those protocols that are already supported.

I think we misunderstood.
I was not argueing solely from a FC perspective: I just said that FC do/can work as a concentrator. They already implement 14 protocols. But they lack a way to transport all that data (see my issues).

On the other hand, i would be of great help, if I could plug in my old FlySky-Sensors to an ELRS-PWM rx - maybe at the same time with an S.Port sensor plugges to a different Uart.

So, who would you expect to use the new telem protocol (maybe aside from oXs)???

The first step sould be to fix the issues we actually have with the transport of new sensor types RX -> TX -> Handset.

**Reply by snakescb** [reply]
@gagarinlg , I see your point. We still need to organize a very universal transport from rx->tx->handheld. Sensor->rx should be considered in CRSF as well, true. At the same time, for Sensor->RX many other established standards already exist, in my opinion the vision should be that we can talk to these as well, longterm. But covering these protocol is an isolated topic. If these are supported, rx needs to translate. We need to make sure in the rx->tx->handlheld communication we do not implement something blocking us from talking to other systems

**Reply by gagarinlg** [reply]
> So, who would you expect to use the new telem protocol (maybe aside from oXs)???
> 

Let's just assume that there are companies waiting for it.

> The first step sould be to fix the issues we actually have with the transport of new sensor types RX -> TX -> Handset.

As I said, I would prefer to do it at the same time, to not have to rip everything apart again, because we forgot something when designing the OTA part or the CRSF part of it.

**Reply by wimalopaan** [reply]
> Let's just assume that there are companies waiting for it.

Ok, you surely know more than me ;-) 

From a physical point of view I think a physical device-chain (like IBus) is a bit cumbersome for most of the users (at least for me, a physical bus topology would be better). 
But a physical bus comes with the problem of bus arbitration in the party-identification phase, and I think common stragegies like multiple-access w/ collision-detection maybe too advanced. So, a device chain would be a good compromise as FlySky-IBus does that. Other strategies as FrSky predefined IDs or Hott-Sensorgroups are too limited. Maybe a tree-based algo like 1-wire would be a chance, but that requires a form of UUID for each device.

> As I said, I would prefer to do it at the same time, to not have to rip everything apart again, because we forgot something when designing the OTA part or the CRSF part of it.

Well, I don't see these too parts coupled very tight together ...

**Reply by CapnBry** [reply]
> We still need to organize a very universal transport from rx->tx->handheld

There already is a very universal transport for sending and receiving data from the RX->TX->handset, it's called MSP. CRSF encapsulated MSP works and can transfer arbitrary data so currently any sensor could be supported. However, you'd have to write something on the handset to take the MSP and generate the handset sensors from it. The problem is you think that there's more bandwidth than there is, at least in the case of ExpressLRS where standard telemetry rates are sub 100 **bits** per second. Even operating the connection in symmetric mode (1:2 telemetry) there is very little effective bandwidth.

The OTA can't be thought of as a transparent pipe for everything to dump into because there's no flow control and the receiver can't buffer everything you send-- you just keep sending more and more data. Even with ELRS Airport running a dedicated second OTA, the bandwidth is just absolutely terrible.

**Comment by schugabe** [comment]
This solution would be very flexible but I see one issue: The available bandwidth between rx/tx is not known and can change during flight (= packages are lost for some reason).

So what should be done in that case? 

ELRS currently has one buffer slot for each type of crsf sensor data and always stores the last received frame and sends the data stored in the buffer. That works really good since it ensures that always fresh data is sent and no sensor is permanently dropped since each has a dedicated buffer slot.

If you add a very flexible system + large number of supported values this would require larger queues or some form of dynamic queue that could lead to sensor starvation in certain cases.

It would also push some of the complexity to the user to choose the correct tlm bandwidth for the connected number of sensors and some form of configurable load shedding to prioritize certain types of sensors.

So it sounds like a elegant solution but increases the complexity quite a bit and that raises the question if that kind of complexity is needed to solve the issue for (most) applications. I would lean more towards the solution of adding more crsf frame definitions (like temperature, second lipo voltage) that could be easily integrated in the existing solutions and keep the simple concept we currently have.

**Reply by snakescb** [reply]
Does not sound good enough to me TBH. Yes bandwith is limitied of course, and needs to be taken into consideration. It adds complexity yes, but the number of supported datatypes is probably not important, it can be just integer based (some sensors may need 1 byte only, others 2, other 4), data is interpreted on the handset. We should maintain a list of all registered sensor id's in the rx memory, together with the latest value reported by this sensor. And then a logic which ensures, that each ID get's some share of the available bandwith. We can give higher priority to system relevant sensors if we want. The less sensors there are, the more frequent the communication happens.

**Reply by wimalopaan** [reply]
> If you add a very flexible system + large number of supported values this would require larger queues or some form of dynamic queue that could lead to sensor starvation in certain cases.

Looks like this is the "real problem" if it is one in the first place. This has to be solved (before the 15th telem protocol is invented).

**Comment by CapnBry** [comment]
This is already an overwhelming pile of chaos for me so I am just going to lay out some more things to consider

* Any sort of ID system must be stable. That is, an ID used to represent a sensor to the handset must remain fixed across reboots of the handset / sensor / receiver / model / etc.
* A single wire bus system can not use physical positioning to determine its ID, everything is on the same wire and nothing knows how far away it is from the receiver. 
* The system must support a standard full duplex UART-based physical layer regardless of if it has its own preferential physical layer. Remember this is for CRSF, so it should be able to run on a single CRSF-enabled pair of RX/TX pads.
* Some devices do report an ID, but sensors should be able to be supported even if they do not supply their own ID.
* The handset can not be expected to download a giant table every time a connection is established. Any sensor discovery must be by user-initiated request, and expect that the control link will be degraded in performance (perhaps severely) during sensor discovery.
* Not everyone flies a single EdgeTX model connecting to a single physical model. Where possible the telemetry system should be functional across physical models with the same reporting devices in it.
* Bandwidth is not unlimited. Items should be able to be reduced down to small packages, and a provision for combining sensors into a larger unit for transfer is needed (i.e. not sending RPM in one packet, then temp in another packet, then current in another packet).
* Grouping of sensors should be able to happen both in the remote sender as well as OTA. You wouldn't want to send like latitude then half a second later send longitude because these values together form one value and can not be separated.
* Most users do not want to have to micromanage their own telemetry scheduling with some sort of confusing priority / deadline / interval. Sensors also can not manage this either, as most the things we're going to be hooking up here have closed firmwares so we'll get what we get. The system will also be generic so it doesn't know what kind of data is being transferred so it can not perform its own QoS. This is a bit of a difficult problem without foisting it on the link system (ELRS/TBS)  to provide the user a UI to manage it.

**Reply by wimalopaan** [reply]
> * A single wire bus system can not use physical positioning to determine its ID, everything is on the same wire and nothing knows how far away it is from the receiver.

I think the term 3-wire (or single wire without power) does not mean that all devices are connected to the very same wire. A device-chain was suggested, and that is how flysky works.

> * The system must support a standard full duplex UART-based physical layer regardless of if it has its own preferential physical layer. Remember this is for CRSF, so it should be able to run on a single CRSF-enabled pair of RX/TX pads.

Why? We are talking about something new.

**Reply by CapnBry** [reply]
> I think the term 3-wire (or single wire without power) does not mean that all devices are connected to the very same wire.

Jeti is something people want to support and it uses 3 wire except on a literal single wire for data. That's why it is important to make this obvious when discussing this. However, I do think discussing the physical layer should be done AFTER the protocol is established because having two conversations while one is already going to be a massive long discussion is just a nightmare.

> Why? We are talking about something new.

Because a flight controller that has new sensors it wants to expose doesn't need to have another wire soldered to the receiver to get those sensors supported.

**Reply by wimalopaan** [reply]
> Jeti is something people want to support and it uses 3 wire except on a literal single wire for data. That's why it is important to make this obvious when discussing this.

Jeti is the same concept as S.Port, SBus2, Hott: it is a simple half-duplex request-response protocol. FlySky/IBus is something different as I mentioned before and as @gagarinlg metioned in the top post.

**Reply by wimalopaan** [reply]
> Because a flight controller that has new sensors it wants to expose doesn't need to have another wire soldered to the receiver to get those sensors supported.

Can't follow ;-) What do you have in mind?

**Reply by CapnBry** [reply]
> Can't follow ;-) What do you have in mind?

I'm not sure how to make that plainer? If an flight controller wanted to report _any of the values you all are so excited to see on your handset_ then a new physical architecture would require adding another wire to the flight controller which would be unacceptable. This **must** be able to pass over an existing UART running the CRSF protocol (other physical architectures notwithstanding).

**Reply by wimalopaan** [reply]
Ok, I see. You have a very limited point of view of FC in general.

So, you suggest connecting the FC as well as additional devices in parallel via the new protocol to a single-wire bus? Then the proposed solution of @gagarinlg with a device-chain would also be a no-go, because this would even require an additional UART oder at least a bypass circuitry (which also would require a modified FC).

Ok, cool. Looks like a real task then ;-)

**Reply by CapnBry** [reply]
I have no idea what you're talking about? I'm saying if a flight controller wants to send this telemetry protocol, is MUST NOT require another wire going to the receiver, because it already has 2 wires for sending CRSF to the receiver that should carry this protocol as well.

I'm not talking about devices beyond the flight controller, just FC to RX, although by the transitive property those devices could also speak CRSF to deliver their telemetry to the FC.

**Reply by wimalopaan** [reply]
> I'm saying if a flight controller wants to send this telemetry protocol, is MUST NOT require another wire going to the receiver, because it already has 2 wires for sending CRSF to the receiver that should carry this protocol as well.

Sorry, but I did not say that.

**Reply by gagarinlg** [reply]
> I have no idea what you're talking about? I'm saying if a flight controller wants to send this telemetry protocol, is MUST NOT require another wire going to the receiver, because it already has 2 wires for sending CRSF to the receiver that should carry this protocol as well.
> 
> I'm not talking about devices beyond the flight controller, just FC to RX, although by the transitive property those devices could also speak CRSF to deliver their telemetry to the FC.

The CRSF Protocol has to be extended anyways for the communication between controller and TX module.
The same messages can be used between RX and FC. The RX might have to do some ID mapping, if it has multiple sensor interfaces, but thats it.

Before anyone can start CRSF protocol extensions the requirements have to be in a state where they do not change every 5 seconds.

The intention is NOT to add anotejhr wire between RX and FC.

@CapnBry can you write down what you would like to change on my first message?

**Comment by sp3swj** [comment]
"temperature telemetry"  - I searched a lot how to send temperature in proper way, Only found this discussion :

https://github.com/rotorflight/rotorflight-firmware/issues/81#issuecomment-2160375732

Is any standardised method defined to send temperature for ESC and motors? 
Maybe for battery can be added temperature too ?
